### PR TITLE
Added ar:// FIX #6104

### LINKED ISSUE #6104: Support ar:// to load archive files like .lib or .a

### PR DESCRIPTION
**Pull request by xarkes**
It works with basic .a I did not test with .lib etc.
2017-08-12-105646_1919x429_scrot

### CODE REVIEW

**Comment by xarkes** [comment]
I forgot to add it to meson build

File: libr/io/p/io_ar.c [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		return true;
+	}
+	return false;
+}
+
+static RIODesc *r_io_ar_open(RIO *io, const char *file, int rw, int mode) {
+	RIODesc *res = NULL;
+	char *arname = strdup (file);
+	arname = strstr (arname, "://");
+	char *filename = strstr(arname + 3, "//");
```
Comment (radare) [inline libr/io/p/io_ar.c]: Arname heap ptr is lost here. Use a different pointer and free the arname at the end

File: libr/io/p/io_ar.c [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+	if (!filename) {
+		eprintf("Wrong filename: use ar://lib.a//filename\n");
```
Comment (radare) [inline libr/io/p/io_ar.c]: Missing space before (

File: libr/io/p/io_ar.c [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		return res;
+	}
+
+	arname[filename - arname] = 0;
```
Comment (radare) [inline libr/io/p/io_ar.c]: Just *filename=0 dont make complex ptr arithmwtics when u already have the pointer

File: libr/io/p/io_ar.c [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	arname = arname + 3;
```
Comment (radare) [inline libr/io/p/io_ar.c]: Use +=

File: libr/io/p/io_ar.c [hunk 5]
```diff
@@ -0,0 +1,102 @@
+/* radare - LGPLv3 - Copyright 2017 - xarkes */
+#include <r_io.h>
+#include <r_lib.h>
+#include <r_util.h>
+#include <r_cons.h>
+#include "ar.h"
+
+static bool r_io_ar_plugin_open(RIO *io, const char *file, bool many) {
+	if (!strncmp("ar://", file, 5) || !strncmp("lib://", file, 6)) {
```
Comment (radare) [inline libr/io/p/io_ar.c]: Remove the if, just return this

Missing space before (

File: libr/io/p/io_ar.c [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		return -1;
+	}
+	arinfo = fd->data;
+	return ar_close (arinfo);
+}
+
+RIOPlugin r_io_plugin_ar = {
+	.name = "ar",
+	.desc = "Open ar/lib files [ar|lib]://[file//path]",
+	.license = "GPLv3",
```
Comment (radare) [inline libr/io/p/io_ar.c]: Why gpl?

File: libr/io/p/io_ar.c [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	.open = r_io_ar_open,
+	.write = r_io_ar_write,
+	.read = r_io_ar_read,
+	.close = r_io_ar_close,
+	.lseek = r_io_ar_lseek,
+	.check = r_io_ar_plugin_open
+};
+
+#ifndef CORELIB
+struct r_lib_struct_t radare_plugin = {
```
Comment (radare) [inline libr/io/p/io_ar.c]: Use the typedef not the struct here. This is RLibStruct

File: libr/io/p/io_ar.c [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	arinfo = fd->data;
+	return ar_close (arinfo);
+}
+
+RIOPlugin r_io_plugin_ar = {
+	.name = "ar",
+	.desc = "Open ar/lib files [ar|lib]://[file//path]",
+	.license = "GPLv3",
+	.open = r_io_ar_open,
```
Comment (radare) [inline libr/io/p/io_ar.c]: Handle open._many like zip:/: does

File: shlr/ar/Makefile [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+OFILES=ar.o
+
+all: ${LIBAR}
+
+%.o: %.c
+	$(CC) $(CFLAGS) $(LDFLAGS) -c $< -o $@
+
+${LIBAR}: $(OFILES)
+	rm -f ${LIBAR}
+	${AR} q ${LIBAR} ${OFILES}
```
Comment (radare) [inline shlr/ar/Makefile]: Use $() instead of ${}

File: shlr/ar/ar.c [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+		/* "//" denotes the start of a filename table */
+		//if (*buffer == '/' && *(buffer + 1) != '/') {
+		/* File identifier */
+		r = fread (buffer + 2, 1, 14, f);
+		if (curfile) free (curfile);
+		curfile = strdup (buffer);
+		tmp = strchr (curfile, '/');
+		*tmp = '\0';
+		if (r != 14) goto end;
+		/* File timestamp */
```
Comment (radare) [inline shlr/ar/ar.c]: Missing braces in every single line of this function

File: shlr/ar/ar.c [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+		r = fread (buffer, 1, 2, f);
+		if (strncmp (buffer, AR_FILE_HEADER_END, 2)) goto end;
+		/* File content */
+		offset = ftell (f);
+		if (fseek (f, filesize, SEEK_CUR)) goto end;
+		//}
+		if (!strcmp (curfile, filename)) break;
+	}
+	free (buffer);
+	desc = malloc (sizeof (RIOArFileObj));
```
Comment (radare) [inline shlr/ar/ar.c]: Null chk

File: shlr/ar/ar.c [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+	RIOArFileObj *desc = NULL;
+	char *curfile = NULL;
+	char *tmp = NULL;
+
+	if (!f) {
+		r_sys_perror (__FUNCTION__);
+		return desc;
+	}
+
+	/* Read magic header */
```
Comment (radare) [inline shlr/ar/ar.c]: Nullchk

File: shlr/ar/ar.c [hunk 13]
```diff
@@ -0,0 +1,100 @@
+/* radare - LGPLv3 - Copyright 2017 - xarkes */
+#include <stdio.h>
+#include <r_util.h>
+#include "ar.h"
+
+const char *AR_MAGIC_HEADER = "!<arch>\n";
+const char *AR_FILE_HEADER_END = "`\n";
+
+RIOArFileObj *ar_open_file(const char *arname, const char *filename) {
+	FILE *f = fopen (arname, "r");
```
Comment (radare) [inline shlr/ar/ar.c]: Use RBuffer. This way u dont depend on libc or filesystem to make this work.

File: shlr/ar/ar.c [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+	if (!f) {
+		r_sys_perror (__FUNCTION__);
+		return desc;
+	}
+
+	/* Read magic header */
+	char *buffer = calloc (1, 1024);
+	r = fread (buffer, 1, 8, f);
+	if (!r) goto end;
```
Comment (radare) [inline shlr/ar/ar.c]: Missing vraces

File: shlr/ar/ar.c [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+	/* Read lookup table */
+	while (true) {
+		r = fread (buffer, 1, 2, f);
+		if (!r) break;
+		/* "//" denotes the start of a filename table */
+		//if (*buffer == '/' && *(buffer + 1) != '/') {
+		/* File identifier */
+		r = fread (buffer + 2, 1, 14, f);
+		if (curfile) free (curfile);
```
Comment (radare) [inline shlr/ar/ar.c]: Unnecessary if. Free(NULL) is totally fine

File: shlr/ar/ar.c (hunk 11 above)
Comment (radare) [inline shlr/ar/ar.c]: Use calloc

File: shlr/ar/ar.c [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+	desc->offset = offset;
+	desc->size = filesize;
+	return desc;
+end:
+	free (buffer);
+	fclose (f);
+	return desc;
+}
+
+int ar_close(RIOArFileObj *arinfo) {
```
Comment (radare) [inline shlr/ar/ar.c]: Use R_API, to make that public

File: shlr/ar/ar.c [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+	return fclose (arinfo->file);
+}
+
+int ar_read_at(RIOArFileObj *arinfo, ut64 off, void *buf, int count) {
+	ar_set_offset(arinfo, off);
+	return fread (buf, 1, count, arinfo->file);
+}
+
+int ar_write_at(RIOArFileObj *arinfo, ut64 off, void *buf, int count) {
+	ar_set_offset(arinfo, off);
```
Comment (radare) [inline shlr/ar/ar.c]: Missing space

File: shlr/ar/ar.h [hunk 18]
```diff
@@ -0,0 +1,18 @@
+/* radare - LGPLv3 - Copyright 2017 - xarkes */
+#ifndef _AR_H
+#define _AR_H
+
+typedef struct RIOArFileObj {
+	FILE *file;
+	char *name;
```
Comment (radare) [inline shlr/ar/ar.h]: Just use RBuffer that abstracts this

File: libr/io/p/io_ar.c (hunk 6 above)
Comment (xarkes) [inline libr/io/p/io_ar.c]: I don't know I don't care I just copy pasted

**Comment by radare** [comment]
Use LGPL, mit or bsd. But gpl is conflicting with lgpl and its a pain for packagers

**Comment by radare** [comment]
please update the pr

**Comment by radare** [comment]
duuuuuude

**Comment by xarkes** [comment]
Yeah one sec

**Comment by radare** [comment]
the second has passed

**Comment by XVilka** [comment]
@xarkes ping?

**Comment by xarkes** [comment]
Sorry I'll push this afternoon

**Pull request merged**